linera-core/src/client.rs

```diff
@@ -364,13 +365,73 @@ where
+                            // TODO(#687): skip notifications that are not useful.
+                            let mut client = this.lock().await;
+                            match &notification.reason {
+                                Reason::NewMessage { origin, height } => {
+                                    if let Err(e) = client.find_received_certificates_from_validator(name, node).await {
+                                        error!("Fail to process notification: {e}");
+                                        return false;
+                                    }
+                                    let Ok(info) = client.node_client.local_chain_info(origin.sender).await else {
+                                        error!("Fail to read local node after notification");
+                                        return false;
+                                    };
+                                    client.update_from_info(&info);
+                                    if info.next_block_height <= *height {
+                                        error!("Fail to synchronize new message after notification");
+                                        return false;
+                                    }
+                                }
+                                Reason::NewBlock { height } => {
+                                    match client.node_client
+                                        .try_synchronize_chain_state_from(name, node, chain_id)
+                                        .await
+                                    {
+                                        Ok(()) => {
+                                            let Ok(info) = client.node_client.local_chain_info(chain_id).await else {
+                                                error!("Fail to read local node after notification");
+                                                return false;
+                                            };
+                                            client.update_from_info(&info);
+                                            if info.next_block_height <= *height {
+                                                error!("Fail to synchronize new block after notification");
+                                                return false;
+                                            }
+                                        }
+                                        Err(e) => {
+                                            error!("Fail to process notification: {e}");
+                                            return false;
+                                        }
+                                    }
+                                }
+                            }
+                            // Keep the notification.
+                            true
+                        })
+                    });
+                    streams.push(stream);
                 }
                 Err(e) => {
```

Comment:
We could `continue` early in case of an error, then the `Ok(stream)` case would have less indentation.

Reply:
I made a separate function